app/models/concerns/unparsed_run.rb

```diff
@@ -201,5 +199,10 @@ def write_segment_histories(segs)
 
       SegmentHistory.import(histories)
     end
+
+    def extract_best_times(segments, timer)
+      seconds = segments.map(&:"#{timer}_best").sum
+      { seconds: seconds, milliseconds: seconds * 1000 }
+    end
```

Comment:
This change is innocuous by itself, but combined with a second or third future innocuous change it can become dangerous! If a future caller behaves poorly like say
```ruby
extract_best_times(segments, params[:timing])
```
then it can open up a way for users to call unallowed methods on `Segment`s by adding `?timing=anything` to their URL -- and we might not realize it because each piece of code on its own looks relatively safe.

One possible solution is to use [`Segment#shortest_duration_ms`][1], which can take a timing (`Run::REAL` / `Run::GAME`) as an argument and behaves ok if it receives something else.

PS: In this repo "timer" refers to the program (LiveSplit etc.) and "timing" refers to the type of time-passing (realtime, gametime).

[1]: https://github.com/glacials/splits-io/blob/master/app/models/segment.rb#L34-L41

Reply:
Goot thinking 👍 I'm still looking into this one though - I just learned that what I called the parameter `segments` here is actually an array of `Split`s, which as far as I can see are a thin wrapper around the parsed hash generated by `Parser`. So I see two alternatives:

* To add a method similar to `Segment#shortest_duration_ms` to `Split`
* To do a `reload` after calling `segments.import` [here](https://github.com/julioolvr/splits-io/blob/18cc412cc8456b43d1909c56ae6c383bfa47faca/app/models/concerns/unparsed_run.rb#L158). That will make the segments available from `Run#segments` and I can then use `Segment#shortest_duration_ms`.

I'm leaning towards the latter but I'm not sure yet, let me know what you think but in any case I'll think about it some more with my morning coffee ☕️

Reply:
Aha, I forgot about `Split`! I say do the latter -- `Split` used to serve a bigger purpose before we switched to livesplit-core for parsing, and now I'm not sure what I want it to be: thin wrapper with no logic, or completely gone. Either way, the second sounds best!

You may even be able to do it without hard reloading if `write_segments` can be modified to return the `Segment`s it imported.

Reply:
Alright, I ended up going with a `reload`, since `activerecord-import` doesn't seem to return instances of the records it saved (which makes sense since they don't seem to instantiate those in memory). I could've done something like finding all the segments by the ids returned by `segments.import` but I feel like it would be pretty much the same as doing `reload` and the latter sounds simpler. That being said I think it's objective so I'm open to changing it.

Reply:
The import call itself takes a bunch of `Segment`s as arguments, are those reusable as return values?

Reply:
🤦‍♂️

Yup, that's right. I'll use those, rename things a bit and send a new commit 👍

Reply:
There! I only renamed `segments` to `splits` since I know have the `segments` variable which holds the `Segment` instances in `#parse_into_db`. It might make sense to actually use those `Segment` instances once they are generated (instead of mixing both during the rest of the method) but I wanted to keep the changes to a minimum for now.